goas/cl/cl/src/balance.rs

```diff
@@ -1,164 +1,156 @@
-}
+        for input in ptx.inputs.iter() {
+            balance.insert_negative(input.note.unit, input.note.value);
+        }
+
+        for output in ptx.outputs.iter() {
+            balance.insert_positive(output.note.unit, output.note.value);
+        }
+
+        balance.clear_zeros();
+
+        balance
+    }
 
-#[cfg(test)]
-mod test {
-    use super::*;
-    use crate::note::unit_point;
-
-    #[test]
-    fn test_balance_zero_unitless() {
-        // Zero is the same across all units
-        let (nmo, eth) = (unit_point("NMO"), unit_point("ETH"));
-
-        let mut rng = rand::thread_rng();
-        let b = BalanceWitness::random(&mut rng);
-        assert_eq!(
-            b.commit([&NoteWitness::basic(0, nmo)], []),
-            b.commit([&NoteWitness::basic(0, eth)], []),
-        );
+    pub fn insert_positive(&mut self, unit: Unit, value: Value) {
+        let mut found = false;
+        for unit_bal in self.balances.iter_mut() {
+            if unit_bal.unit == unit {
+                found = true;
+                unit_bal.pos += value;
+                break;
+            }
+        }
+
+        if !found {
+            self.balances.push(UnitBalance {
+                unit,
+                pos: value,
+                neg: 0,
+            });
+        }
     }
 
-    #[test]
-    fn test_balance_blinding() {
-        // balances are blinded
-        let nmo = unit_point("NMO");
+    pub fn insert_negative(&mut self, unit: Unit, value: Value) {
+        let mut found = false;
+        for unit_bal in self.balances.iter_mut() {
+            if unit_bal.unit == unit {
+                found = true;
+                unit_bal.neg += value;
+                break;
+            }
+        }
+
+        if !found {
+            self.balances.push(UnitBalance {
+                unit,
+                pos: 0,
+                neg: value,
+            });
+        }
+    }
 
-        let r_a = Scalar::from(12u32);
-        let r_b = Scalar::from(8u32);
-        let bal_a = BalanceWitness::new(r_a);
-        let bal_b = BalanceWitness::new(r_b);
+    pub fn clear_zeros(&mut self) {
+        let mut i = 0usize;
+        while i < self.balances.len() {
+            if self.balances[i].is_zero() {
+                self.balances.swap_remove(i);
+                // don't increment `i` since the last element has been swapped into the
+                // `i`'th place
+            } else {
+                i += 1;
+            }
+        }
+    }
 
-        let note = NoteWitness::basic(10, nmo);
+    pub fn combine(balances: impl IntoIterator<Item = Self>, blinding: [u8; 16]) -> Self {
+        let mut combined = BalanceWitness::zero(blinding);
 
-        let a = bal_a.commit([&note], []);
-        let b = bal_b.commit([&note], []);
+        for balance in balances {
+            for unit_bal in balance.balances.iter() {
+                if unit_bal.pos > unit_bal.neg {
```

Comment:
is it an invariant that only one of the negative or positive value is set?

Reply:
no, it's not, but we only have helper functions for inserting a negative or positive u64.

Reply:
yeah sorry misunderstood